---
title: "Exploring my GPX4i genome-wide CRISPR screen data"
author: "Tanja Leont'Ãªva"
number-sections: true
format: 
  html:
    toc: true
---


```{r}
#| label: global_options
#| echo: false
knitr::opts_chunk$set(cache = TRUE)
# options(error = recover)
```

# Setup
Load and attach required packages.

```{r}
#| label: setup
#| message: false
library("tibble")
library("readr")    # for data import
library("dplyr")
library("tidyr")
library("ggplot2")
library("GGally")   # for the pairs plot
library("DESeq2")   # for the library size normalization
```

Read the count table

```{r}
#| label: read
x = readr::read_delim("counts.count.txt.gz", col_types = c("cciiiiiii"))
```

# Quality assessment, visualization and library size normalization

Get a first overview: pairwise scatterplot.
The list `panel_args` provides the plot options (smaller points, diagonal red line).
We also use `mutate` to apply a shifted log transformation, i.e. `log10(n+1)`.

```{r}
#| label: ggpairs-raw
#| fig-width: 10
#| fig-height: 10
panel_args = list(continuous = 
  function(data, mapping, ...) {
    ggplot(data = data, mapping = mapping) +
      geom_point(size = 0.1, alpha = 0.5) +
      geom_abline(intercept = 0, slope = 1, color = "red") +
      coord_fixed()
  })
ggpairs(mutate(x[, 3:9], across(where(is.numeric), function(n) log10(n+1))), 
        upper = panel_args, 
        lower = panel_args)
```

Overall, looks good! 
Note that in the scatterplots between T0 and DMSO conditions, for both cell lines, we see a lot of decreased counts - i.e. sgRNAs with a viability / proliferation effect, irrespective of the drug.

Now normalize for sequencing depth using the `DESeq2` method, and apply the shifted log transformation. We omit the plasmid library.

```{r}
#| label: normalize
xn = x
sf = DESeq2::estimateSizeFactorsForMatrix(as.matrix(x[, 3:8]))
round(sf, 3)
for (i in names(sf)) 
  xn[, i] = log10(xn[, i] / sf[i] + 1)
```

```{r}
#| label: ggpairs-norm
#| fig-width: 10
#| fig-height: 10
ggpairs(xn[, 3:8], upper = panel_args, lower = panel_args)
```

# Select interesting sgRNAs ("hits")

Select sgRNAs that decrase by less than 1.5-fold from T0 to DMSO (`is_not_viability_hit`), 
and decrease or increase by more than 3-fold from DMSO to ML210, in both cell lines.
Note that we are working on a logarithmic scale, so division is subtraction.

```{r}
#| label: select
theta1 = log10( 1.5 )
theta2 = log10( 3 )

xn = mutate(xn,
  is_not_viability_hit = (JEKO_T0 - JEKO_DMSO < theta1) & (MAVER_T0 - MAVER_DMSO < theta1),
  up   = is_not_viability_hit & (JEKO_DMSO - JEKO_ML210 >  theta2) & (MAVER_DMSO - MAVER_ML210 >  theta2),   
  down = is_not_viability_hit & (JEKO_DMSO - JEKO_ML210 < -theta2) & (MAVER_DMSO - MAVER_ML210 < -theta2)
)
table(xn$is_not_viability_hit)
table(xn$up)
table(xn$down)
```

## Select interesting genes: 

The sgRNA library is such that most target genes have 5 sgRNAs:
(The "gene" with 612 sgRNAs is `r names(which(table(x$Gene)>100))`.)

```{r}
#| label: numsgrnas
table(table(xn$Gene))
```

So let's find genes with >=3 sgRNAs among the `up` and `down` sgRNA sets.

```{r}
#| label: tabup+down
tabup   = table(xn$Gene[xn$up])
tabdown = table(xn$Gene[xn$down])

names(which(tabup>=3))
names(which(tabdown>=3))
```

# Plot exemplary data

For this, we pivot `xn` to long format, and decompose `name` into the additional columns `cellline` and `condition`, resulting in `xnl`.

```{r}
#| label: pivot
xnl = tidyr::pivot_longer(xn, cols = JEKO_DMSO:MAVER_T0) |>
  mutate(cellline  = stringr::str_split_fixed(name, "_", 2)[,1],
         condition = stringr::str_split_fixed(name, "_", 2)[,2] |> 
           ordered(levels = c("T0", "DMSO", "ML210"))
         )
xnl    
```
... and define the helper function `plotGene`:

```{r}
#| label: plotagene
#| fig-width: 7
#| fig-height: 3
plotGene = function(gene)
  dplyr::filter(xnl, Gene == gene) |>
    ggplot(aes(x = condition, y = value, col = sgRNA)) +
      geom_point() + geom_line(aes(x = as.integer(condition))) +
      facet_grid(cols = vars(cellline))
plotGene("ABCG1")
plotGene("FUT8")
```

# To Do

Play with the above threshold parameters; send the list of genes in, e.g., `up` to ChatGPT or similar, with a prompt asking for genes whose role is somewhat "surprising" for the biological process we aim to study here (GPX4 inhibition).